v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/uniformsplitter/range/BoundarySplitterFactory.java

```diff
@@ -0,0 +1,153 @@
+      ImmutableMap.of(
+          Integer.class,
+              (BoundarySplitter<Integer>)
+                  (start, end, partitionColumn, boundaryTypeMapper, processContext) ->
+                      splitIntegers(start, end),
+          Long.class,
+              (BoundarySplitter<Long>)
+                  (start, end, partitionColumn, boundaryTypeMapper, processContext) ->
+                      splitLongs(start, end),
+          BigInteger.class,
+              (BoundarySplitter<BigInteger>)
+                  (start, end, partitionColumn, boundaryTypeMapper, processContext) ->
+                      splitBigIntegers(start, end),
+          String.class, (BoundarySplitter<String>) BoundarySplitterFactory::splitStrings);
+
+  /**
+   * Creates {@link BoundarySplitter BoundarySplitter&lt;T&gt;} for pass class {@code c} such that
+   * {@code T.getClass == c}.
+   *
+   * @param c Class for the type of {@link BoundarySplitter}
+   * @return {@link BoundarySplitter BoundarySplitter&lt;T&gt;}
+   * @throws UnsupportedOperationException if BoundarySplitter is not yet implemented.
+   */
+  public static <T extends Serializable> BoundarySplitter<T> create(Class c) {
+
+    BoundarySplitter<T> splitter = (BoundarySplitter<T>) splittermap.get(c);
+    if (splitter == null) {
+      throw new UnsupportedOperationException("Range Splitter not implemented for class " + c);
+    }
+    return splitter;
+  }
+
+  private BoundarySplitterFactory() {}
+
+  private static Integer splitIntegers(Integer start, Integer end) {
+    if (start == null && end == null) {
+      return null;
+    }
+    if (start == null) {
+      start = Integer.MIN_VALUE;
+    }
+    if (end == null) {
+      end = Integer.MIN_VALUE;
+    }
+
+    // This is same as (start + (end - start)/2) but that the width calculation can overflow if
+    // start is INTEGER.MIN and end is INTEGER.MAX.
+    return (start & end) + ((start ^ end) >> 1);
+  }
+
+  private static BigInteger splitBigIntegers(BigInteger start, BigInteger end) {
+    BigInteger low;
+    BigInteger high;
+    if (start == null && end == null) {
+      return null;
+    }
+    if (start == null) {
+      start = BigInteger.ZERO;
+    }
+    if (end == null) {
+      end = BigInteger.ZERO;
+    }
+
+    if (start.compareTo(end) <= 0) {
+      low = start;
+      high = end;
+    } else {
+      high = start;
+      low = end;
+    }
+    return low.add((high.subtract(low)).divide(BigInteger.TWO));
+  }
+
+  private static Long splitLongs(Long start, Long end) {
+    Long low;
+    Long high;
+
+    if (start == null && end == null) {
+      return null;
+    }
+    if (start == null) {
+      start = Long.MIN_VALUE;
+    }
+    if (end == null) {
+      end = Long.MIN_VALUE;
+    }
+
+    // This is same as (start + (end - start)/2) but that the width calculation can overflow if
+    // start is LONG.MIN and end is LONG.MAX.
+    return (start & end) + ((start ^ end) >> 1);
```

Comment:
this logic is not very clear. Please add an example if possible.

Reply:
Sure. I will add. 



At a high level it's the average between start and end. But,

(Context on going this route) - a general long column say if the user has chosen to put in random integers , it could get values from too low on the negative spectrum to too high on positive spectrum.

if we do (start + end)/2 - it overflows if the sum goes too high , both start and end are near Long.Max.
If this was unsigned , we could do with start + (end - start)/2 - the `half-width` method. But that overflows when start and end are on 2 ends of positive and negative long.
for example Long.Min + (Long.Max - Long.Min)/2 will now work as (Long.Max. - Long.Min) will overflow (even if we half them individually, it comes dangerously close to 2 * (Long.Max/2)

Sum of 2 numbers can be written as 

`(start ^ end)  + ((start & end)<<1)` basically xor of bits and add the carries.
So average simply becomes (start + end)/2
`(start ^ end)>>1  + ((start & end))`

I have added UT at all the corners of Long.MAX and Long.Min to verify that this does not overflow and does give an average rounded down (rounding down consistently is also important pice for us)

Reply:
Added a comment explaining this in more detail.